### PR TITLE
Fix issue 325 by adding algo for equals()

### PR DESCRIPTION
**Pull request by nainar**
@shans could you take a look, took a stab at writing up the algo for .equals()

### CODE REVIEW

File: css-typed-om/Overview.bs [hunk 1]
```diff
@@ -784,10 +782,20 @@ The following are the arithmetic operations you can perform on dimensions:
...
 </div>
 
-<div class=issue>
-    Define {{CSSNumericValue/equals()}}.
+<div algorithm="CSSNumericValue.equals(...|value|)">
+    The <dfn for=CSSNumericValue method>equals(...|value|)</dfn> method,
+    when called on a {{CSSNumericValue}} |this|,
+    must perform the following steps:
+
+    1. If the |value| item is the same type as the {{CSSNumericValue}} the method is being called on continue to step 2, else return false.
```
Comment (shans) [inline css-typed-om/Overview.bs]: You don't need to continue on to step 2 as this happens automatically. Maybe rewrite as 'If |value| is of different type to |this|, return false.'

File: css-typed-om/Overview.bs [hunk 2]
```diff
@@ -784,10 +782,20 @@ The following are the arithmetic operations you can perform on dimensions:
         is set to |this|.
 </div>
 
-<div class=issue>
-    Define {{CSSNumericValue/equals()}}.
+<div algorithm="CSSNumericValue.equals(...|value|)">
+    The <dfn for=CSSNumericValue method>equals(...|value|)</dfn> method,
+    when called on a {{CSSNumericValue}} |this|,
+    must perform the following steps:
+
```
Comment (shans) [inline css-typed-om/Overview.bs]: Will need to either explicitly iterate over |value| (change to values?) or just restrict this to take a single value.

File: css-typed-om/Overview.bs (hunk 1 above)
Comment (shans) [inline css-typed-om/Overview.bs]: Also we need to define what 'different type' means.

File: css-typed-om/Overview.bs [hunk 3]
```diff
@@ -784,10 +782,20 @@ The following are the arithmetic operations you can perform on dimensions:
...
+    1. If the |value| item is the same type as the {{CSSNumericValue}} the method is being called on continue to step 2, else return false.
+
+    2. If the |value| item has the same type as the {{CSSNumericValue}} continue to the next step, else return false.
```
Comment (shans) [inline css-typed-om/Overview.bs]: Isn't this the same as step 1?

File: css-typed-om/Overview.bs [hunk 4]
```diff
@@ -784,10 +782,20 @@ The following are the arithmetic operations you can perform on dimensions:
...
+
+    3. If the |value| item has the same unit as the {{CSSNumericValue}} return true, else return false.
+
```
Comment (shans) [inline css-typed-om/Overview.bs]: Also need to handle calc values, etc.

File: css-typed-om/Overview.bs (hunk 3 above)
Comment (nainar) [inline css-typed-om/Overview.bs]: This was supposed to be class.

File: css-typed-om/Overview.bs [hunk 5]
```diff
@@ -784,10 +782,20 @@ The following are the arithmetic operations you can perform on dimensions:
         is set to |this|.
 </div>
 
-<div class=issue>
-    Define {{CSSNumericValue/equals()}}.
+<div algorithm="CSSNumericValue.equals(...|value|)">
+    The <dfn for=CSSNumericValue method>equals(...|value|)</dfn> method,
```
Comment (tabatkins) [inline css-typed-om/Overview.bs]: Note that .equals() is variadic; you have to check that *all* the values are equal to each other/this.

File: css-typed-om/Overview.bs (hunk 1 above)
Comment (tabatkins) [inline css-typed-om/Overview.bs]: * Just reverse this conditional and return false. No need to explicitly say to go to the next step; that's implied by the list.  
* Be clear what you mean by "type" - you use the exact same word in the second step, but clearly mean something different.  I suspect here you're talking about the interface type.
* "the CSSNumericValue the method is being called on" is already stored in the `|this|` variable, so you just refer to it that way.

File: css-typed-om/Overview.bs (hunk 3 above)
Comment (tabatkins) [inline css-typed-om/Overview.bs]: * Same concerns as about the structure; reverse the conditional and just return false.
* I think you mean "value" here? Refer to `|this|’s {{CSSUnitValue/value}} internal slot` then (and same for |value|).

File: css-typed-om/Overview.bs [hunk 6]
```diff
@@ -784,10 +782,20 @@ The following are the arithmetic operations you can perform on dimensions:
...
+    when called on a {{CSSNumericValue}} |this|,
+    must perform the following steps:
+
+    1. If the |value| item is the same type as the {{CSSNumericValue}} the method is being called on continue to step 2, else return false.
+
+    2. If the |value| item has the same type as the {{CSSNumericValue}} continue to the next step, else return false.
+
+    3. If the |value| item has the same unit as the {{CSSNumericValue}} return true, else return false.
```
Comment (tabatkins) [inline css-typed-om/Overview.bs]: Use `{{CSSUnitValue/unit}} internal slot`.

File: css-typed-om/Overview.bs [hunk 7]
```diff
@@ -784,10 +782,20 @@ The following are the arithmetic operations you can perform on dimensions:
         is set to |this|.
 </div>
 
-<div class=issue>
-    Define {{CSSNumericValue/equals()}}.
+<div algorithm="CSSNumericValue.equals(...|value|)">
+    The <dfn for=CSSNumericValue method>equals(...|value|)</dfn> method,
+    when called on a {{CSSNumericValue}} |this|,
+    must perform the following steps:
```
Comment (tabatkins) [inline css-typed-om/Overview.bs]: The algo claims to be defined for CSSNumericValue, but the steps assume that you're looking at a CSSUnitValue.  It needs to work for all the CSSMathValue subtypes too.

**Comment by tabatkins** [comment]
Whoops, I still had this page up from earlier, and it hadn't updated with Shane's comments. Sorry for doubling up. ^_^

File: css-typed-om/Overview.bs (hunk 2 above)
Comment (tabatkins) [inline css-typed-om/Overview.bs]: And by "or" Shane means "do the first thing", because all the other methods take multiple values too.

**Comment by nainar** [comment]
I suck at Github but I have added a new commit. Hopefully that addresses both @shans and @tabatkins comments.

File: css-typed-om/Overview.bs (hunk 7 above)
Comment (nainar) [inline css-typed-om/Overview.bs]: Done in commit 3

File: css-typed-om/Overview.bs (hunk 2 above)
Comment (nainar) [inline css-typed-om/Overview.bs]: Had done the first thing. Commit 3 should make equals accept an array.

File: css-typed-om/Overview.bs (hunk 4 above)
Comment (nainar) [inline css-typed-om/Overview.bs]: Done in commit 3

File: css-typed-om/Overview.bs (hunk 1 above)
Comment (nainar) [inline css-typed-om/Overview.bs]: Done in commit 3

File: css-typed-om/Overview.bs (hunk 3 above)
Comment (nainar) [inline css-typed-om/Overview.bs]: Done in commit 3

File: css-typed-om/Overview.bs (hunk 6 above)
Comment (nainar) [inline css-typed-om/Overview.bs]: Done in commit 3

File: css-typed-om/Overview.bs [hunk 8]
```diff
@@ -784,10 +782,29 @@ The following are the arithmetic operations you can perform on dimensions:
...
+
+    1. Replace each [=list/item=] of |values|
+        with the result of [=rectifying a numberish value=] for the [=list/item=].
+
+    2. Let |thisAndValues| be |this| prepended to |values|.
+
+    1. [=list/If all=] |item|s of |thisAndValues| do not belong to the same class return false. Example: all items in |thisAndValues| must be {{CSSUnitValue}}.
+
+    2. [=list/If all=] |item|s in |thisAndValues| are {{CSSUnitValue}s and have the same {{CSSUnitValue/value}} return true.
```
Comment (tabatkins) [inline css-typed-om/Overview.bs]: unit *and* value ^_^

File: css-typed-om/Overview.bs [hunk 9]
```diff
@@ -784,10 +782,29 @@ The following are the arithmetic operations you can perform on dimensions:
...
+
+    4. =list/If all=] |item|s in |thisAndValues| are {{CSSMathValue}s :
```
Comment (tabatkins) [inline css-typed-om/Overview.bs]: Missing `[`

File: css-typed-om/Overview.bs [hunk 10]
```diff
@@ -784,10 +782,29 @@ The following are the arithmetic operations you can perform on dimensions:
...
+        1. If the {{CSSMathValue/operator}} are not equal for all {{CSSMathValues}} in |thisAndValues| return false.
```
Comment (tabatkins) [inline css-typed-om/Overview.bs]: This is already covered by "same class" earlier, so you can drop this.

File: css-typed-om/Overview.bs [hunk 11]
```diff
@@ -784,10 +782,29 @@ The following are the arithmetic operations you can perform on dimensions:
...
+        2. If |thisAndValues| are all either {{CSSMathSum}}, {{CSSMathProduct}}, {{CSSMathMin}} or {{CSSMathMax}}, iterate over the {{CSSNumericArray}}s on |thisAndValues|,
+        	1. For each {{CSSUnitValue}} in each of the {{CSSNumericArray}}s go back to step 1 to check that the {{CSSUnitValue}}s on all items in |thisAndValue| are equal.
+    	3. If |this| and |value| are both either {{CSSMathNegate}} or {{CSSMathInvert}}:
```
Comment (tabatkins) [inline css-typed-om/Overview.bs]: Need to handle multiples.

**Comment by nainar** [comment]
@tabatkins took a stab at equals take a look?

File: css-typed-om/Overview.bs (hunk 8 above)
Comment (nainar) [inline css-typed-om/Overview.bs]: Done

File: css-typed-om/Overview.bs (hunk 9 above)
Comment (nainar) [inline css-typed-om/Overview.bs]: Done

File: css-typed-om/Overview.bs (hunk 10 above)
Comment (nainar) [inline css-typed-om/Overview.bs]: Done

File: css-typed-om/Overview.bs (hunk 11 above)
Comment (nainar) [inline css-typed-om/Overview.bs]: Do you mean CSSMathProduct? Cause that is dealt with in the above step. Or do you mean the case where values in |thisAndValues| are not all of the same type? If the latter in that case we should return a false right?

**Pull request merged**